crates/provider-wit-bindgen/src/lib.rs

```diff
@@ -1808,15 +2184,130 @@ fn convert_wit_type(t: &wit_parser::Type, cfg: &ProviderBindgenConfig) -> anyhow
+            let ok_ty = if let Some(ty_id) = ok {
+                convert_wit_type(ty_id, cfg)?
+            } else {
+                quote::quote!(())
+            };
+            let err_ty = if let Some(ty_id) = err {
+                convert_wit_type(ty_id, cfg)?
             } else {
-                bail!("failed to parse wit type def for type {t:#?}");
+                quote::quote!(())
+            };
+            Ok(quote::quote!(Result<#ok_ty, #err_ty>))
+        }
+        // Nested type case (List<...>)
+        TypeDefKind::List(ty_id) => {
+            let ty = convert_wit_type(ty_id, cfg)?;
+            Ok(quote::quote!(Vec<#ty>))
+        }
+        // Nested type case (owned data)
+        TypeDefKind::Handle(Handle::Own(ty_idx)) => {
+            let ty_def = cfg
+                .wit_bindgen_cfg
+                .as_ref()
+                .context("missing WIT bindgen cfg resolver")?
+                .resolve
+                .types
+                .get(*ty_idx)
+                .context("failed to find type with given ID in WIT resolver")?;
+            convert_wit_typedef(ty_def, cfg)
+        }
+        // Nested type case (borrowed data)
+        TypeDefKind::Handle(Handle::Borrow(ty_idx)) => {
+            let ty_def = cfg
+                .wit_bindgen_cfg
+                .as_ref()
+                .context("missing WIT bindgen cfg resolver")?
+                .resolve
+                .types
+                .get(*ty_idx)
+                .context("failed to find type with given ID in WIT resolver")?;
+            convert_wit_typedef(ty_def, cfg)
+        }
+        // Nested type case (Tuple)
+        TypeDefKind::Tuple(Tuple { types }) => {
+            let mut tuple_types = TokenStream::new();
+            for (idx, tokens) in types
+                .iter()
+                .map(|t| convert_wit_type(t, cfg))
+                .collect::<anyhow::Result<Vec<TokenStream>>>()
+                .context("failed to parse all types in Tuple w/ types {types:?}")?
+                .iter()
+                .enumerate()
+            {
+                tuple_types.append_all(quote::quote!(#tokens));
+                if idx != types.len() - 1 {
+                    tuple_types.append(TokenTree::Punct(Punct::new(
+                        ',',
+                        proc_macro2::Spacing::Alone,
+                    )));
+                }
             }
+            Ok(
+                proc_macro2::Group::new(proc_macro2::Delimiter::Parenthesis, tuple_types)
+                    .to_token_stream(),
+            )
+        }
+        TypeDefKind::Stream(Stream { element, .. }) => {
+            let element_ty = convert_wit_type(&element.context("missing type for stream")?, cfg)?;
+            Ok(quote::quote!(impl Stream<Item=#element_ty>))
+        }
+        // Nested types that come through can run through
+        // there's a potential for a cycle here, but it's, but it's unlikely
```

Comment:
Nit: typo in "but it's, but it's"

Reply:
Will fix -- yeah looks like I repeated myself somehow there